src/gui/guiScene.cpp

```diff
@@ -0,0 +1,239 @@
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "guiScene.h"
+
+#include <IGUIEnvironment.h>
+#include <SViewFrustum.h>
+#include <IAnimatedMeshSceneNode.h>
+#include <ILightSceneNode.h>
+
+static const v2f rotation_bound{60.f, 300.f};
+
+GUIScene::GUIScene(gui::IGUIEnvironment *env, scene::ISceneManager *smgr,
+		   gui::IGUIElement *parent, core::recti rect, s32 id)
+	: IGUIElement(gui::EGUIET_ELEMENT, env, parent, id, rect)
+{
+	m_driver = env->getVideoDriver();
+	m_smgr = smgr->createNewSceneManager(false);
+
+	m_cam = m_smgr->addCameraSceneNode(0, v3f(0.f, 0.f, -100.f), v3f(0.f));
+	m_cam->setFOV(30.f * core::DEGTORAD);
+
+	scene::ILightSceneNode *light = m_smgr->addLightSceneNode(m_cam);
+	light->setRadius(1000.f);
+
+	m_smgr->getParameters()->setAttribute(scene::ALLOW_ZWRITE_ON_TRANSPARENT, true);
+}
+
+scene::IAnimatedMeshSceneNode *GUIScene::setMesh(scene::IAnimatedMesh *mesh)
+{
+	if (!mesh) {
+		if (m_mesh) {
+			m_mesh->remove();
+			m_mesh = nullptr;
+		}
+		return m_mesh;
+	} else if (m_mesh) {
+		m_mesh->remove();
+		m_mesh = nullptr;
+	}
+
+	m_mesh = m_smgr->addAnimatedMeshSceneNode(mesh);
+	m_mesh->setPosition(-m_mesh->getBoundingBox().getCenter());
+	m_mesh->animateJoints();
+
+	return m_mesh;
+}
+
+void GUIScene::setTexture(u32 idx, video::ITexture *texture)
+{
+	video::SMaterial &material = m_mesh->getMaterial(idx);
+	material.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
+	material.MaterialTypeParam = 0.5f;
+	material.TextureLayer[0].Texture = texture;
+	material.setFlag(video::EMF_LIGHTING, false);
+	material.setFlag(video::EMF_FOG_ENABLE, true);
+	material.setFlag(video::EMF_BILINEAR_FILTER, false);
+	material.setFlag(video::EMF_BACK_FACE_CULLING, false);
+}
+
+void GUIScene::draw()
+{
+	core::rect<s32> oldViewPort = m_driver->getViewPort();
+	m_driver->setViewPort(getAbsoluteClippingRect());
+	core::recti borderRect = Environment->getRootGUIElement()->getAbsoluteClippingRect();
+
+	if (m_bgcolor != 0) {
+		Environment->getSkin()->draw3DSunkenPane(
+			this, m_bgcolor, false, true, borderRect, 0);
+	}
+
+	core::dimension2d<s32> size = getAbsoluteClippingRect().getSize();
+	m_smgr->getActiveCamera()->setAspectRatio((f32)size.Width / (f32)size.Height);
+
+	if (!m_target) {
+		updateCamera(m_smgr->addEmptySceneNode());
+		rotateCamera(v3f(0.f));
+		m_cam->bindTargetAndRotation(true);
+	}
+
+	cameraLoop();
+
+	if (m_autorotate)
+		rotateCamera(v3f(0.f, -0.5f, 0.f));
```

Comment:
You can already specify a rotation speed here, so why not expose it to the element definition? It's next to no effort, but provides more options to modders.